prusti-interface/src/environment/mod.rs

```diff
@@ -305,7 +305,19 @@ impl<'tcx> Environment<'tcx> {
     pub fn type_is_copy(&self, ty: ty::Ty<'tcx>) -> bool {
         let copy_trait = self.tcx.lang_items().copy_trait();
         if let Some(copy_trait_def_id) = copy_trait {
-            self.type_implements_trait(ty, copy_trait_def_id)
+            match ty.kind() {
+                ty::TyKind::Ref(_, _, mir::Mutability::Not) => {
+                    // Shared references are copy.
+                    true
+                }
+                ty::TyKind::Array(_, _) | ty::TyKind::Slice(_) => {
+                    // Arrays and slices are not copy.
+                    false
+                }
+                _ => {
+                    self.type_implements_trait(ty, copy_trait_def_id)
+                }
```

Comment:
Why this weird check? Is our notion of `copy` different from types that implement Rust's `Copy` trait?

Reply:
Our implementation of `type_implements_trait` is wrong. I should add a comment explaining that.